File: lista3/cpp/GF.cpp
// GF.cpp

#include "GF.h"

GF::GF() : value(0) {}

GF::GF(long value) : value((value % PRIME_NUMBER + PRIME_NUMBER) % PRIME_NUMBER) {}

long GF::characteristic() {
    return PRIME_NUMBER;
}

long GF::getValue() {
    return value;
}

bool GF::operator ==(const GF& other) {
    return value == other.value;
}

bool GF::operator !=(const GF& other) {
    return !(*this == other);
}

bool GF::operator <=(const GF& other) {
    return value <= other.value;
}

bool GF::operator >=(const GF& other) {
    return value >= other.value;
}

bool GF::operator <(const GF& other) {
    return value < other.value;
}

bool GF::operator >(const GF& other) {
    return value > other.value;
}

void GF::operator +=(const GF& other) {
    value = (value + other.value) % PRIME_NUMBER;
}

void GF::operator -=(const GF& other) {
    value = (value - other.value + PRIME_NUMBER) % PRIME_NUMBER;
}

void GF::operator *=(const GF& other) {
    value = (value * other.value) % PRIME_NUMBER;
}

void GF::operator /=(const GF& other) {
    if (other.value == 0)
        throw std::invalid_argument("Division by zero");
    *this *= other.inverse();
}

GF GF::inverse() const {
    if (value == 0)
        throw std::invalid_argument("Inverse of zero does not exist");

    long long int base = value;
    long long int exponent = PRIME_NUMBER - 2;
    long long int result = 1;
    while (exponent > 0) {
        if (exponent % 2 == 1) {
            result = (result * base) % PRIME_NUMBER;
        }
        base = (base * base) % PRIME_NUMBER;
        exponent /= 2;
    }
    return GF(static_cast<int>(result));
}


GF operator +(const GF& l, const GF& r) {
    GF result = l;
    result += r;
    return result;
}

GF operator -(const GF& l, const GF& r) {
    GF result = l;
    result -= r;
    return result;
}

GF operator *(const GF& l, const GF& r) {
    GF result = l;
    result *= r;
    return result;
}

GF operator /(const GF& l, const GF& r) {
    GF result = l;
    result /= r;
    return result;
}

std::ostream& operator <<(std::ostream& os, const GF& gf) {
    os << gf.value;
    return os;
}

std::istream& operator >>(std::istream& is, GF& gf) {
    long temp;
    is >> temp;
    gf = GF(temp);
    return is;
}

File: lista3/python/GF.py
# GF.py

class GF:
    PRIME_NUMBER = 1234567891

    def __init__(self, value=0):
        self.value = (value % self.PRIME_NUMBER + self.PRIME_NUMBER) % self.PRIME_NUMBER

    @staticmethod
    def characteristic(self):
        return self.PRIME_NUMBER

    def __eq__(self, other):
        return self.value == other.value

    def __ne__(self, other):
        return not self == other

    def __le__(self, other):
        return self.value <= other.value

    def __ge__(self, other):
        return self.value >= other.value

    def __lt__(self, other):
        return self.value < other.value

    def __gt__(self, other):
        return self.value > other.value

    def __add__(self, other):
        return GF((self.value + other.value) % self.PRIME_NUMBER)

    def __sub__(self, other):
        return GF((self.value - other.value + self.PRIME_NUMBER) % self.PRIME_NUMBER)

    def __mul__(self, other):
        return GF((self.value * other.value) % self.PRIME_NUMBER)

    def __truediv__(self, other):
        if other.value == 0:
            raise ValueError("Division by zero")
        return self * other.inverse()

    def __iadd__(self, other):
        self.value = (self.value + other.value) % self.PRIME_NUMBER
        return self

    def __isub__(self, other):
        self.value = (self.value - other.value + self.PRIME_NUMBER) % self.PRIME_NUMBER
        return self

    def __imul__(self, other):
        self.value = (self.value * other.value) % self.PRIME_NUMBER
        return self

    def __itruediv__(self, other):
        if other.value == 0:
            raise ValueError("Division by zero")
        self *= other.inverse()
        return self

    def inverse(self):
        if self.value == 0:
            raise ValueError("Inverse of zero does not exist")

        base = self.value
        exponent = self.PRIME_NUMBER - 2
        result = 1
        while exponent > 0:
            if exponent % 2 == 1:
                result = (result * base) % self.PRIME_NUMBER
            base = (base * base) % self.PRIME_NUMBER
            exponent //= 2
        return GF(result)

    def __str__(self):
        return str(self.value)

    def __repr__(self):
        return f"GF({self.value})"

    @staticmethod
    def from_input(value):
        return GF(value)
